File: core/src/main/java/io/github/frikit/krandom/generator/base/IntGenerator.java
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package io.github.frikit.krandom.generator.base;

/**
 * Generates random {@link Integer} values.
 *
 * <p>Default range: [{@code Integer.MIN_VALUE}, {@code Integer.MAX_VALUE}) — the full 32-bit range
 * minus the maximum value itself (exclusive upper bound).
 *
 * <pre>{@code
 *   int any     = new IntGenerator().generate();
 *   int positive = new IntGenerator(1, Integer.MAX_VALUE).generate();
 *   int roll     = new IntGenerator(1, 7).generate();  // die roll [1..6]
 * }</pre>
 */
public final class IntGenerator extends AbstractBoundedGenerator<Integer> {

    public IntGenerator() {
        super(Integer.MIN_VALUE, Integer.MAX_VALUE, null);
    }

    public IntGenerator(int min, int max) {
        super(min, max, null);
    }

    public IntGenerator(int min, int max, long seed) {
        super(min, max, seed);
    }

    /**
     * Generate an int in the half-open range [{@code min}, {@code max}).
     *
     * @throws IllegalArgumentException if {@code min == max}
     */
    @Override
    public Integer generate(Integer min, Integer max) {
        validate(min, max);
        int lo = lo(min, max);
        int hi = hi(min, max);
        return random.nextInt(lo, hi);
    }
}

File: core/src/main/kotlin/org/github/krandom/common/generators/IntGenerator.kt
/*
 * Copyright (c) 2026 krandom contributors
 *
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
package org.github.krandom.common.generators

import org.github.krandom.common.RangedTypeGenerator
import java.security.SecureRandom
import java.util.Random

/**
 * Generates random [Int] values.
 *
 * - [generate()] generates within the configured [start, end) range (defaults: full Int range).
 * - [generate(start, end)] uses a half-open range [start, end): start is inclusive, end is exclusive.
 */
class IntGenerator(
    private val start: Int = Int.MIN_VALUE,
    private val end: Int = Int.MAX_VALUE,
    seed: Long? = null
) : RangedTypeGenerator<Int> {

    private val random: Random = if (seed != null) Random(seed) else SecureRandom()

    override fun generate(): Int = generate(start, end)

    override fun generate(start: Int, end: Int): Int {
        require(start != end) { "start ($start) and end ($end) must differ" }
        val lo = minOf(start, end)
        val hi = maxOf(start, end)
        return random.ints(1, lo, hi).findFirst().orElse(lo)
    }
}
